Implement string conversions

akx opened the pull request:
This implements the three string transformations described in #436:

> **Transform**
> Similar to https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html#intrinsic-string-manipulation-types
>
> ```yaml
> NEW_VAR: 
>   transform: [uppercase, lowercase, capitalize]
>   source: $VAR
> ```

(with the keyword `convert` as discussed in the comments/review)

Here's my actual use-case (slightly anonymized):

```yaml
id: foo
language: python
rule:
  pattern: ObjectStatus.$A
  regex: ObjectStatus.(.*)_statuses
transform:
  B:
    replace:
      source: $A
      replace: '(.*)_statuses'
      by: '${1}_STATUSES'
  CHANGED:
    convert:
      source: $B
      convert: uppercase
fix:
  $CHANGED
```

This turns `ObjectStatus.foo_statuses` to `FOO_OBJECT_STATUSES` everywhere.

HerringtonDarkholme commented:
Hi @akx , thanks for the pull request!

I have tried Rust's regex crate and it looks like there is no way to transform string cases via regex.

While I do think it is useful for case transformation. The API needs some discussion. 
Having two `transformation`s does not look right to me.

HerringtonDarkholme commented:
I think the word `convert` will be better. So the ending API can be 

```yaml
transform:
  NEW_VAR:
    source: '$A'
    convert: uppercase
```

akx commented:
Sure, I can rename the field (and I guess `StringTransformation` to `StringConversion`).

akx commented:
Okay, it's `convert` now, but I'm not quite sure how to wrangle `pub enum Transformation` to always accept a `convert` – it'd now be

```
transform:
  NEW_VAR:
    convert:
      source: '$A'
      convert: uppercase
```

and

```
transform:
  CHANGED:
    replace:
      source: $A
      replace: '(.*)_statuses'
      by: '${1}_STATUSES'
      convert: uppercase
```

– I suppose we'd want

```
transform:
  NEW_VAR:
    source: '$A'
    convert: uppercase
```

and

```
transform:
  CHANGED:
    convert: uppercase
    replace:
      source: $A
      replace: '(.*)_statuses'
      by: '${1}_STATUSES'
```

but I'll look into that!

HerringtonDarkholme commented on crates/config/src/transform.rs:
```diff
@@ -66,11 +62,34 @@ fn resolve_char(opt: &Option<i32>, dft: i32, len: i32) -> usize {
+  source: String,
+  replace: String,
+  by: String,
+  convert: Option<StringConversion>,
```
Can we also remove the convert here? Composing transformation can be achieved now by 

```
transform:
  REPLACE:
    replace:
      source: '$A'
      replace: regex
      by: xxx
  CASE:
      convert:
          source: '$REPLACE'
          convert: uppercase
```

Indeed it is more verbose but it is more consistent with other transformations.
And we still have the leeway to add `convert` in future without breaking change.

akx replied:
Can do. That also answers my other question – I won't move `source` and `convert` to the "main" Transform level :)

HerringtonDarkholme reviewed (approved):
LGTM!

HerringtonDarkholme commented:
I will try to release this tomorrow! Thanks for your contribution!

akx commented:
@HerringtonDarkholme Oops, yeah, sorry! It's there now.

HerringtonDarkholme commented:
Hi @akx thanks for your contribution!

I have also slightly modified the API to 

```yml
NEW_VAR:
  convert:
    letterCase: uppercase
    source: $A
```

so there is no two same convert and the goal of `convert` reads explicitly as `letterCase`.

I will release it in 0.12.1!

HerringtonDarkholme commented:
Released in 0.12.1!

akx commented:
Sure, that makes sense!

Thank you for accepting my contribution and thank you for the project in the first place!
